File: guests/haskell/textarea.hs
{-# LANGUAGE OverloadedStrings #-}

-- The textarea scene, Haskell port — guests/rust/textarea.rs,
-- tools/scenes/textarea.steps.

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

lineTally :: Text -> Text
lineTally text
  | T.null text = "0 lines"
  | otherwise = T.pack (show (length (T.lines text)) ++ " lines")

main :: IO ()
main = kayaMain $ \app -> do
  (lineCount, editor, clearBtn) <- buildTx app $ do
    window primary [WTitle "textarea"]
    lineCount <- signal (T.pack "0 lines")

    -- Realized here because the handlers below need their handles.
    editor <- textarea
    clearBtn <- button "clear"
    root <- column [] [pure editor, labelBound lineCount, pure clearBtn]
    mount root
    return (lineCount, editor, clearBtn)

  onChange app editor $ \text ->
    submitTx app $ writeSignal lineCount (lineTally text)
  onClick app clearBtn $
    submitTx app $ do
      clearWidget editor
      focusWidget editor

File: guests/rust/textarea.rs
//! The textarea conformance scene (tools/scenes/textarea.steps): a newline
//! rides the text both ways, and `clear` echoes text_changed("").

#[derive(Clone)]
enum Msg {
    Edited(String),
    Clear,
}

fn count(text: &str) -> String {
    if text.is_empty() {
        "0 lines".to_string()
    } else {
        format!("{} lines", text.lines().count())
    }
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let (lines, editor) = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW).title("textarea");
        let lines = tx.signal("0 lines");
        let (root, editor) = tx
            .column(|tx| {
                let editor = tx.textarea().id();
                msgs.on_change(editor, Msg::Edited);
                tx.label(lines);
                let clear = tx.button("clear").id();
                msgs.on_click(clear, Msg::Clear);
                editor
            })
            .into_parts();
        tx.mount(root);
        (lines, editor)
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::Edited(text) => {
                ctx.apply(|tx| {
                    tx.write(lines, count(&text));
                });
            }
            Msg::Clear => {
                ctx.apply(|tx| {
                    tx.clear(editor);
                    tx.focus(editor);
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}
